celements-reference/src/main/java/org/xwiki/model/reference/EntityReference.java

```diff
@@ -0,0 +1,206 @@
+
+  public EntityReference(String name, EntityType type) {
+    this(name, type, null);
+  }
+
+  public EntityReference(String name, EntityType type, EntityReference parent) {
+    setName(name);
+    setType(type);
+    setParent(parent);
+  }
+
+  /**
+   * @exception IllegalArgumentException
+   *              if the passed name is null or empty
+   */
+  public void setName(String name) {
+    if (StringUtils.isEmpty(name)) {
+      throw new IllegalArgumentException("An Entity Reference name cannot be null or empty");
+    }
+    this.name = name;
+  }
+
+  public String getName() {
+    return this.name;
+  }
+
+  public void setParent(EntityReference parent) {
+    this.parent = parent;
+    if (parent != null) {
+      parent.setChild(this);
+    }
+  }
+
+  public EntityReference getParent() {
+    return this.parent;
+  }
+
+  public void setChild(EntityReference child) {
+    this.child = child;
+  }
+
+  public EntityReference getChild() {
+    return this.child;
+  }
+
+  /**
+   * @exception IllegalArgumentException
+   *              if the passed type is null
+   */
+  public void setType(EntityType type) {
+    if (type == null) {
+      throw new IllegalArgumentException("An Entity Reference type cannot be null");
+    }
+    this.type = type;
+  }
+
+  public EntityType getType() {
+    return this.type;
+  }
+
+  public EntityReference getRoot() {
+    EntityReference reference = this;
+    while (reference.getParent() != null) {
+      reference = reference.getParent();
+    }
+    return reference;
+  }
+
+  public EntityReference extractReference(EntityType type) {
+    EntityReference reference = this;
+
+    while ((reference != null) && (reference.getType() != type)) {
+      reference = reference.getParent();
+    }
+
+    return reference;
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see java.lang.Object#toString()
+   */
+  @Override
+  public String toString() {
+    return "name = [" + getName() + "], type = [" + getType() + "], parent = [" + getParent() + "]";
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see java.lang.Object#equals(Object)
+   */
+  @Override
+  public boolean equals(Object obj) {
+    boolean equals = false;
+
+    if (obj == this) {
+      equals = true;
+    } else if (obj instanceof EntityReference) {
+      EntityReference entityReference = (EntityReference) obj;
+
+      equals = (entityReference.getName() == null ? getName() == null
+          : entityReference.getName().equals(getName()))
+          && (entityReference.getParent() == null ? getParent() == null
+              : entityReference.getParent().equals(
+                  getParent()))
+          && (entityReference.getType() == null ? getType() == null
+              : entityReference.getType().equals(
+                  getType()));
+    }
+
+    return equals;
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see Object#hashCode()
+   */
+  @Override
+  public int hashCode() {
+    return toString().hashCode();
```

Comment:
better use Objects.hash

Reply:
see comment above

Reply:
I think that in the moment when we fork code, we make it to ower own code. This code therefore needs a proper review and quality improvments to level our code quality expectations. If done it for you, because you did not oppose the suggestion.